packages/discovery/src/discovery/provider/DiscoveryProvider.ts

```diff
@@ -56,13 +59,56 @@ export class DiscoveryProvider {
     return Bytes.fromHex(result)
   }
 
-  async getLogs(
+  public async getLogs(
     address: EthereumAddress,
     topics: (string | string[])[],
     fromBlock: number,
     toBlock: number,
   ): Promise<providers.Log[]> {
-    return this.provider.getLogs({
+    if (fromBlock > toBlock) {
+      throw new Error(
+        `fromBlock (${fromBlock}) can't be bigger than toBlock (${toBlock})`,
+      )
+    }
+
+    if (this.getLogsMaxRange === undefined) {
+      return await this.getLogsBatch(address, topics, fromBlock, toBlock)
+    }
+
+    // To support efficient caching, we divide the requested blocks range into
+    // sequential boundaries of `maxRange` size, e.g [0,10k-1], [10k, 20k-1], ...
+    // Otherwise ranges would depend on `fromBlock` and even small change to it
+    // would make the previous cache useless.
+
+    // Let's start with the deployment block number if it's higher than fromBlock
+    const { blockNumber: deploymentBlockNumber } =
+      await this.getDeploymentInfo(address)
+
+    const maxRange = this.getLogsMaxRange
+    const allLogs: providers.Log[][] = []
+
+    let start = Math.max(fromBlock, deploymentBlockNumber)
+    let end
+    do {
+      const curBoundaryStart = Math.floor(start / maxRange) * maxRange
+      const curBoundaryEnd = curBoundaryStart + maxRange - 1 // getLogs 'to' is inclusive!
+      end = Math.min(curBoundaryEnd, toBlock)
+      const logs = await this.getLogsBatch(address, topics, start, end)
+      allLogs.push(logs)
+      start = end + 1
+    } while (start <= toBlock)
```

Comment:
Have you tried `await `Promise.all()` all the batches? Wouldn't it speed up the process?

Reply:
It wouldn't work here, because usually we either have a cache, so we just make one query, or we don't have cache and we make hundreds of queries for batches. Running hundred requests in parallel (Promise.all) would go over limits of connections, overwhelm the node etc.

I'm aware we can add optimizations here (limited parallel workers, etc.) but in practice it wouldn't make a lot of difference. There are other places in discovery were parallelization would be more urgent. So keeping this implementation as is for now.